x/programs/program/reentrancy.go

```diff
@@ -0,0 +1,57 @@
+// Copyright (C) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package program
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/ava-labs/avalanchego/ids"
+)
+
+var ErrReentrancy = errors.New("reentrant call not allowed")
+
+type reentrancyMap map[ids.ID]map[string]bool
+
+// ReentrancyGuard keeps track of called methods during execution.
+type ReentrancyGuard struct {
+	// In the future, every call context might need its own reentrancyMap
+	// with fine grain locking across call contexts if we want to parellellize program calls.
+	m reentrancyMap
+}
+
+func NewReentrancyGuard() *ReentrancyGuard {
+	return &ReentrancyGuard{
+		m: make(reentrancyMap),
+	}
+}
+
+// Reset resets the reentrancy map.
+func (r *ReentrancyGuard) Reset() {
+	r.m = make(reentrancyMap)
+}
+
+// Note, you can bypass this check by calling a different method for the same program which call the intended no reenter method.
+// ex. Program A has method no_reenter and reenter. Reenter calls method no_reenter.
+// If an external call wants to call no_reenter, it can call reenter instead and will bypass the reentrancy check.
+// The second call wont be an external program call, but it will be a reentrant call and is not handled.
+
+// Enter returns if we have never entered this function before, errors otherwise.
+// If the function has not been entered before, we set the visited flag.
+// We can return an error because only non-reentrant functions will call this method.
+func (r *ReentrancyGuard) Enter(id ids.ID, fn string) error {
```

Comment:
Could we make this a little more abstract?  It might be nice if there were a reentrancyGuard interface with the param of Enter either being generic or just an interface so we could have different functionality without having to add a ton of plumbing.